### PR TITLE
Add Dynamic fields to Snowflake Connection 

### PR DESCRIPTION
**Pull request by sunkickr**
Adds form fields and custom form behavior for the Snowflake connection so it is more obvious to new users what fields need to be filled out. I have divided the connection inputs into three main categories and created three proposals. The code in this PR implements proposal 1. Also the doc_strings for the hook are updated to reflect the params along with helpful information using sphinx directives. 

Fields/inputs Used by the average user

- Conn Id
- Conn Type
- Host
- Schema
- Login
- Password
- account
- database
- region

Fields/inputs not used by user

- Port

Fields/inputs not used by user but reflected in code(not used by average user)

- role
- authenticator
- private_key_field
- session_parameters
- aws_access_key_id
- aws_secret_access_key

## Proposal 1

- Add fields used by the average user, remove fields not used by the users, and allow users to use extras for fields that are not used by average user
- Add placeholders so users have a little more information on fields
    - ex. Login: "your snowflake username"

screenshot:
image

## Proposal 2

- Add all fields used reflected as options in the code except AWS, remove fields not used by the users, allow users to use extras for AWS
- Add placeholders so users have a little more information on fields
    - ex. Login: "your snowflake username"

## Proposal 3

- Add all fields used reflected as options in the code, remove fields not used by the users including extras
- Add placeholders so users have a little more information on fields
    - ex. Login: "your snowflake username"

### CODE REVIEW

**Comment by eladkal** [comment]
We should not suggest to save secrets in extra field if it's not encrypted.
We have the same issue with Salesforce https://github.com/apache/airflow/issues/8766

**Comment by sunkickr** [comment]
@eladkal I agree the easiest solution I see would be too add encrypted fields for all secrets. Is there a better solution I am missing here?

**Comment by sunkickr** [comment]
https://github.com/apache/airflow/blob/b718495e4caecb753742c3eb22919411a715f24a/tests/core/test_providers_manager.py#L81-L82 It appears that this issue as been fixed?

**Comment by kaxil** [comment]
> https://github.com/apache/airflow/issues/12881

yup it's fixed

File: airflow/providers/snowflake/hooks/snowflake.py [hunk 1]
```diff
@@ -58,10 +135,10 @@ def _get_conn_params(self) -> Dict[str, Optional[str]]:
...
             self.snowflake_conn_id  # type: ignore[attr-defined] # pylint: disable=no-member
         )
-        account = conn.extra_dejson.get('account', '')
-        warehouse = conn.extra_dejson.get('warehouse', '')
-        database = conn.extra_dejson.get('database', '')
-        region = conn.extra_dejson.get("region", '')
+        account = conn.extra_dejson.get('extra__snowflake__account', '')
+        warehouse = conn.extra_dejson.get('extra__snowflake__warehouse', '')
+        database = conn.extra_dejson.get('extra__snowflake__database', '')
+        region = conn.extra_dejson.get('extra__snowflake__region', '')
```
Comment (kaxil) [inline airflow/providers/snowflake/hooks/snowflake.py]: This is a breaking change, current users who use `account` / `warehouse` / `database` or `region` won't work

Comment (sunkickr) [reply airflow/providers/snowflake/hooks/snowflake.py]: world this be a proper fix? 
```python
account = conn.extra_dejson.get('extra__snowflake__account', '') \
            or conn.extra_dejson.get('account', '')
```

Comment (kaxil) [reply airflow/providers/snowflake/hooks/snowflake.py]: Something like this

```python
account = conn.extra_dejson.get('extra__snowflake__account', '')
dep_account = conn.extra_dejson.get('account', '')
if dep_account:
    warnings.warn(
        "Passing Snowflake account via extras is deprecated, set the account field instead'",
        DeprecationWarning,
        stacklevel=2,
    )
    account = dep_account

```

Comment (sunkickr) [reply airflow/providers/snowflake/hooks/snowflake.py]: thank you!

Comment (ashb) [reply airflow/providers/snowflake/hooks/snowflake.py]: Hold off on deprecating the old name -- I'd like to see if it's possible to not "leak" the `extra__snowflake__` part -- that's a UI abstraction that has crept in that shouldn't be in the extra JSON.

**Comment by ashb** [comment]
> We should not suggest to save secrets in extra field if it's not encrypted.
> We have the same issue with Salesforce #8766

I don't think there is anything sensitive in the extra field here, is there?

**Comment by sunkickr** [comment]
> > We should not suggest to save secrets in extra field if it's not encrypted.
> > We have the same issue with Salesforce #8766
> 
> I don't think there is anything sensitive in the extra field here, is there @eladkal?
> 
> ```
>         account = conn.extra_dejson.get('extra__snowflake__account', '')
>         warehouse = conn.extra_dejson.get('extra__snowflake__warehouse', '')
>         database = conn.extra_dejson.get('extra__snowflake__database', '')
>         region = conn.extra_dejson.get('extra__snowflake__region', '')
> ```
> 
> Those all seem safe.

This may be the code that he is worried about
https://github.com/apache/airflow/blob/99aab051600715a1ad029ce45a197a8492e5a151/airflow/providers/snowflake/hooks/snowflake.py#L136-L139

File: airflow/providers/snowflake/hooks/snowflake.py (hunk 1 above)
Comment (kaxil) [reply airflow/providers/snowflake/hooks/snowflake.py]: If there is a way that would be good but currently it is used in a some providers:

```
airflow/providers/cncf/kubernetes/hooks/kubernetes.py:47:    - use in cluster configuration by using ``extra__kubernetes__in_cluster`` in connection
airflow/providers/cncf/kubernetes/hooks/kubernetes.py:48:    - use custom config by providing path to the file using ``extra__kubernetes__kube_config_path``
airflow/providers/cncf/kubernetes/hooks/kubernetes.py:50:        ``extra__kubernetes__kube_config`` in connection
airflow/providers/cncf/kubernetes/hooks/kubernetes.py:77:            "extra__kubernetes__in_cluster": BooleanField(lazy_gettext('In cluster configuration')),
airflow/providers/cncf/kubernetes/hooks/kubernetes.py:78:            "extra__kubernetes__kube_config_path": StringField(
airflow/providers/cncf/kubernetes/hooks/kubernetes.py:81:            "extra__kubernetes__kube_config": StringField(
airflow/providers/cncf/kubernetes/hooks/kubernetes.py:84:            "extra__kubernetes__namespace": StringField(
airflow/providers/cncf/kubernetes/hooks/kubernetes.py:108:        in_cluster = extras.get("extra__kubernetes__in_cluster")
airflow/providers/cncf/kubernetes/hooks/kubernetes.py:109:        kubeconfig_path = extras.get("extra__kubernetes__kube_config_path")
airflow/providers/cncf/kubernetes/hooks/kubernetes.py:110:        kubeconfig = extras.get("extra__kubernetes__kube_config")
airflow/providers/cncf/kubernetes/hooks/kubernetes.py:115:                "Invalid connection configuration. Options extra__kubernetes__kube_config_path, "
airflow/providers/cncf/kubernetes/hooks/kubernetes.py:116:                "extra__kubernetes__kube_config, extra__kubernetes__in_cluster are mutually exclusive. "
airflow/providers/cncf/kubernetes/hooks/kubernetes.py:213:        namespace = extras.get("extra__kubernetes__namespace", "default")
airflow/providers/google/ads/hooks/ads.py:134:        secret = secret_conn.extra_dejson["extra__google_cloud_platform__keyfile_dict"]
airflow/providers/google/cloud/hooks/cloud_sql.py:402:GCP_CREDENTIALS_KEY_PATH = "extra__google_cloud_platform__key_path"
airflow/providers/google/cloud/hooks/cloud_sql.py:403:GCP_CREDENTIALS_KEYFILE_DICT = "extra__google_cloud_platform__keyfile_dict"
airflow/providers/google/cloud/hooks/cloud_sql.py:531:            project_id = connection.extra_dejson.get('extra__google_cloud_platform__project')
airflow/providers/google/cloud/hooks/dataprep.py:50:        self._token = extra_dejson.get("extra__dataprep__token")
airflow/providers/google/cloud/hooks/dataprep.py:51:        self._base_url = extra_dejson.get("extra__dataprep__base_url", "https://api.clouddataprep.com")
airflow/providers/google/cloud/operators/cloud_sql.py:1102:                'extra__google_cloud_platform__project'
airflow/providers/google/cloud/utils/credentials_provider.py:63:    extras = "extra__google_cloud_platform"
airflow/providers/google/common/hooks/base_google.py:173:            "extra__google_cloud_platform__project": StringField(
airflow/providers/google/common/hooks/base_google.py:176:            "extra__google_cloud_platform__key_path": StringField(
airflow/providers/google/common/hooks/base_google.py:179:            "extra__google_cloud_platform__keyfile_dict": PasswordField(
airflow/providers/google/common/hooks/base_google.py:182:            "extra__google_cloud_platform__scope": StringField(
airflow/providers/google/common/hooks/base_google.py:185:            "extra__google_cloud_platform__num_retries": IntegerField(
airflow/providers/google/common/hooks/base_google.py:293:        long_f = f'extra__google_cloud_platform__{f}'
airflow/providers/grpc/hooks/grpc.py:63:            "extra__grpc__auth_type": StringField(
airflow/providers/grpc/hooks/grpc.py:66:            "extra__grpc__credential_pem_file": StringField(
airflow/providers/grpc/hooks/grpc.py:69:            "extra__grpc__scopes": StringField(
airflow/providers/grpc/hooks/grpc.py:163:        full_field_name = f'extra__grpc__{field_name}'
airflow/providers/jdbc/hooks/jdbc.py:50:            "extra__jdbc__drv_path": StringField(lazy_gettext('Driver Path'), widget=BS3TextFieldWidget()),
airflow/providers/jdbc/hooks/jdbc.py:51:            "extra__jdbc__drv_clsname": StringField(
airflow/providers/jdbc/hooks/jdbc.py:69:        jdbc_driver_loc: Optional[str] = conn.extra_dejson.get('extra__jdbc__drv_path')
airflow/providers/jdbc/hooks/jdbc.py:70:        jdbc_driver_name: Optional[str] = conn.extra_dejson.get('extra__jdbc__drv_clsname')
airflow/providers/yandex/hooks/yandex.py:49:            "extra__yandexcloud__service_account_json": PasswordField(
airflow/providers/yandex/hooks/yandex.py:56:            "extra__yandexcloud__service_account_json_path": StringField(
airflow/providers/yandex/hooks/yandex.py:63:            "extra__yandexcloud__oauth": PasswordField(
airflow/providers/yandex/hooks/yandex.py:69:            "extra__yandexcloud__folder_id": StringField(
airflow/providers/yandex/hooks/yandex.py:75:            "extra__yandexcloud__public_ssh_key": StringField(
airflow/providers/yandex/hooks/yandex.py:135:        long_f = f'extra__yandexcloud__{field_name}'
```

**Comment by petedejoy** [comment]
So do we think it makes sense to expose the AWS secret as a separate optional field on the form? A bit counter-intuitive, but I don't see any other way to design around it without building a feature that allows the user to encrypt extras.

File: airflow/providers/snowflake/hooks/snowflake.py (hunk 1 above)
Comment (sunkickr) [reply airflow/providers/snowflake/hooks/snowflake.py]: Should I just make sure it works with both names for now? Without the deprecation warning?

Comment (kaxil) [reply airflow/providers/snowflake/hooks/snowflake.py]: hmm sure

**Comment by sunkickr** [comment]
I attempted to rebase my branch onto apache master and it added all commits that have been merged this week.

**Pull request closed without merging**